example-raft-kv/src/network/api.rs

```diff
@@ -20,9 +27,56 @@ use crate::store::ExampleRequest;
  */
 #[post("/write")]
 pub async fn write(app: Data<ExampleApp>, req: Json<ExampleRequest>) -> actix_web::Result<impl Responder> {
-    let request = ClientWriteRequest::new(EntryPayload::Normal(req.0));
-    let response = app.raft.client_write(request).await;
-    Ok(Json(response))
+
+    // Redirect this call to the leader, this is required because a `Candidate`, `Follower` or `Learner` is not allowed to write data to
+    // other nodes. The payload has to wrapped around the type fo expect payload that raft uses. In this case it is real payload and
+    // `EntryPayload::Normal(D)` will be used.
+    let request: ClientWriteRequest<ExampleRequest> = ClientWriteRequest::new(EntryPayload::Normal(req.0.clone()));
+    
+    // Attempt to write the data to the raft and dispatch to the nodes.
+    let response: Result<ClientWriteResponse<ExampleResponse>, ClientWriteError> = app.raft.client_write(request).await;
+
+    match response {
+        Ok(_) => Ok(HttpResponse::Ok().finish()),
+        Err(error) => {
+            match error {
+                ClientWriteError::ForwardToLeader(foward_to_leader) => 
+                    redirect_to_leader(&app, &req, "/write", foward_to_leader).await,
+                _ => Ok(HttpResponse::BadRequest().body("Failed write number"))
+            }
+        }
+    }
+}
+
+async fn redirect_to_leader<T: Serialize>(
+  app: &Data<ExampleApp>,
+  req: &Json<T>,
+  route: &str,
+  foward_to_leader: ForwardToLeader,
+) -> actix_web::Result<HttpResponse> {
+
+  // In a real world application you should not unwrap this, instead return a error related to the missing leader id.
+  let leader_id: NodeId = foward_to_leader.leader_id.unwrap();
+
+  let state_machine = app.store.state_machine.read().await;
+  let nodes = state_machine.nodes.clone();
```

Comment:
The lock should be released ASAP, or this node may be blocked forever since the lock will not be released before the RPC is done.

Reply:
This code below suffies the issue?

```
let nodes = {
  let state_machine = app.store.state_machine.read().await; // Move happens here
  state_machine.nodes.clone() // `state_machine` is released
};
```

Reply:
Looks good :DDD